feat(core): update `Injector.get` and `TestBed.inject` to support object-based flags

AndrewKushnir opened the pull request:
See individual commits.

## PR Type
What kind of change does this PR introduce?

- [x] Feature

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

AndrewKushnir committed:
feat(core): support object-based DI flags in Injector.get()

This commit applies the changes similar to the ones performed for the `inject()` function in https://github.com/angular/angular/pull/46649/commits/df246bb2358794e3aad53870f1554d84bbe390b2.

The `Injector.get` function is updated to use previously added object-based API for options: now the flags argument supports passing an object which configures injection flags.

DEPRECATED:

The bit field signature of `Injector.get()` has been deprecated, in favor of the new options object.

AndrewKushnir committed:
feat(core): support object-based DI flags in TestBed.inject()

This commit applies the changes similar to the ones performed for the `inject()` function in https://github.com/angular/angular/pull/46649/commits/df246bb2358794e3aad53870f1554d84bbe390b2.

The `TestBed.inject` function is updated to use previously added object-based API for options: now the flags argument supports passing an object which configures injection flags.

DEPRECATED:

The bit field signature of `TestBed.inject()` has been deprecated, in favor of the new options object.

AndrewKushnir commented:
[Exploratory Presubmit](http://test/OCL:460085725:BASE:460085755:1657488626784:aa9bac4) (TGP will also be required).

atscott commented on goldens/public-api/core/testing/index.md:
```diff
@@ -172,8 +177,12 @@ export interface TestBedStatic {
-    inject<T>(token: ProviderToken<T>, notFoundValue?: T, flags?: InjectFlags): T;
+    inject<T>(token: ProviderToken<T>, notFoundValue?: T, options?: InjectOptions): T;
     // (undocumented)
+    inject<T>(token: ProviderToken<T>, notFoundValue: null, options?: InjectOptions): T | null;
```
A few questions here:

* Should `Injector.get` have an override like this one? 
* Shouldn't this include `undefined` as well as `null`?
* I think return value `| null` depends on `optional` being explicitly `true`, right?

AndrewKushnir replied:
> Can we add a TODO for v15 then?

Added a note to revisit this in one of the major versions.

> Both? TestBed.inject mimics injector.get. But also Injector.get doesn't have this override at the moment anyways. The previous bullet was requesting to add it.

Added the override that we've discussed.

AndrewKushnir commented:
[TGP #2](https://test.corp.google.com/OCL:460085725:BASE:460631437:1657690975961:2319f4a8).

alxhub commented on packages/core/src/di/injector_compatibility.ts:
```diff
@@ -240,17 +211,22 @@ export function inject<T>(token: ProviderToken<T>, options: InjectOptions): T|nu
+// Converts object-based DI flags (`InjectOptions`) to bit flags (`InjectFlags`).
+export function convertToBitFlags(flags: InjectOptions|InjectFlags|undefined): InjectFlags|
+    undefined {
+  if (typeof flags === 'undefined' || typeof flags === 'number') return flags;
```
```suggestion
  if (typeof flags === 'undefined' || typeof flags === 'number') {
    return flags;
  }
```

alxhub commented on packages/core/src/di/injector_compatibility.ts:
```diff
@@ -240,17 +211,22 @@ export function inject<T>(token: ProviderToken<T>, options: InjectOptions): T|nu
+}
+
+// Converts object-based DI flags (`InjectOptions`) to bit flags (`InjectFlags`).
+export function convertToBitFlags(flags: InjectOptions|InjectFlags|undefined): InjectFlags|
```
```suggestion
export function convertToBitFlags(flags: undefined): undefined;
export function convertToBitFlags(flags: InjectOptions|InjectFlags): InjectFlags;
export function convertToBitFlags(flags: InjectOptions|InjectFlags|undefined): InjectFlags|
```

Setting up the signature like this will allow you to skip the `as InjectFlags` cast later when TS knows that `flags` is defined.

alxhub reviewed (commented):
FYI, I think that modifying the signature of `Injector.get` is a breaking change. We should add the appropriate notice and push this to v15.

AndrewKushnir commented:
Discussed offline: this is a breaking change, so we postpone it for now and will come back to this PR once the `main` branch opens up for v15 changes (after v14.2.0 release). I've applied the necessary labels and added this PR to the `v15-candidates` milestone.

AndrewKushnir commented:
[Presubmit #3](http://test/OCL:473891486:BASE:473891527:1663031040658:38744377).

pkozlowski-opensource commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -3444,6 +3444,117 @@ describe('di', () => {
 
+  describe('injection flags', () => {
+    describe('represented as an options object argument', () => {
+      it('should be able to optionally inject a service', () => {
```
nit: I find the title of this test a bit misleading since we don't inject any service. I would expect something along the lines of "should return null when trying to inject non-existing service with optional flag" - or something like this.

pkozlowski-opensource reviewed (commented):
@AndrewKushnir the change LGTM overall

alxhub commented on packages/core/src/di/injector.ts:
```diff
@@ -45,10 +45,24 @@ export abstract class Injector {
+   * @returns The instance from the injector if defined, otherwise the `notFoundValue`.
+   * @throws When the `notFoundValue` is `undefined` or `Injector.THROW_IF_NOT_FOUND`.
+   */
+  abstract get<T>(token: ProviderToken<T>, notFoundValue?: T, options?: InjectOptions|InjectFlags):
```
We should absolutely extend the type here so that if you pass `undefined, {optional: true}` you get a `T|null` return type. Ditto for `TestBed.inject()`.

AndrewKushnir replied:
Thanks for the feedback @alxhub 👍 I've pushed a fixup commit:
https://github.com/angular/angular/pull/46761/commits/8690b13fcb8813263aedeefef0fbab4c7076a1ce

Could you please take another look?

AndrewKushnir committed:
fixup! feat(core): support object-based DI flags in TestBed.inject()

AndrewKushnir commented:
@alxhub I've updated the API and added a couple tests as we discussed (fixup commit: https://github.com/angular/angular/pull/46761/commits/6159102533dc3f61ba39f0dc5a6d735ccedc2831). Could you please take another look?

AndrewKushnir commented:
[One more TGP before the submission](https://test.corp.google.com/OCL:477065069:BASE:477065121:1664250511881:faede2ec).

jessicajaniuk reviewed (approved):
reviewed-for: public-api

alxhub reviewed (approved):
Reviewed-for: public-api

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api 
Reviewed-for: fw-core

alxhub commented:
This PR was merged into the repository by commit 120555a626d66523e46ce01681b11a0f38e3a40a.
